### PR TITLE
Identify issuer on revocation

### PR DESCRIPTION
**Pull request by cipherboy**
When we attempt to revoke a leaf certificate, we already parse all of
the issuers within the mount (to x509.Certificate) to ensure we don't
accidentally revoke an issuer via the leaf revocation endpoint. We can
reuse this information to associate the issuer (via issuer/subject
comparison and signature checking) to the revoked cert in its revocation
info. This will help OCSP, avoiding the case where the OCSP handler
needs to associate a certificate to its issuer.

`Signed-off-by: Alexander Scheel <alex.scheel@hashicorp.com>`

---

- [x] Wait for #16723 and #16762 to merge.

### CODE REVIEW

File: builtin/logical/pki/crl_test.go [hunk 1]
```diff
@@ -904,6 +910,57 @@ func TestAutoRebuild(t *testing.T) {
 	})
 	require.NoError(t, err)
 
+	// Now, we want to test the issuer identification on revocation. This
+	// only happens as a distinct "step" when CRL building isn't done on
+	// each reovcation. Pull the storage from the cluster and verify the
```
Comment (kitography) [inline builtin/logical/pki/crl_test.go:915]: nit: revocation.

File: builtin/logical/pki/crl_test.go [hunk 2]
```diff
@@ -806,6 +807,7 @@ func TestAutoRebuild(t *testing.T) {
 		LogicalBackends: map[string]logical.Factory{
 			"pki": Factory,
 		},
+		EnableRaw: true,
```
Comment (kitography) [inline builtin/logical/pki/crl_test.go:810]: Can you share what this is?

File: builtin/logical/pki/crl_test.go [hunk 3]
```diff
@@ -904,6 +910,57 @@ func TestAutoRebuild(t *testing.T) {
...
+	var pkiMount string
+	storage := cluster.Cores[0].UnderlyingStorage
+	mounts, err := storage.List(ctx, "logical/")
+	require.NoError(t, err)
+	require.NotEmpty(t, mounts)
+	for _, mount := range mounts {
+		// For whatever reason, OIDC gets provisioned as the first mount,
+		// but I'm not convinced that's a stable list. Let's look inside
+		// each mount until we find a revoked certs folder that we'd expect
+		// if its a real PKI mount. This is because mounts are just UUID
```
Comment (kitography) [inline builtin/logical/pki/crl_test.go:927]: nit: it's

File: builtin/logical/pki/crl_util.go [hunk 4]
```diff
@@ -626,6 +633,20 @@ func buildCRLs(ctx context.Context, b *backend, req *logical.Request, forceNew b
 	return nil
 }
 
+func associateRevokedCertWithIsssuer(revInfo *revocationInfo, revokedCert *x509.Certificate, issuerIDCertMap map[issuerID]*x509.Certificate) bool {
+	for issuerId, issuerCert := range issuerIDCertMap {
+		if bytes.Equal(revokedCert.RawIssuer, issuerCert.RawSubject) {
+			if err := revokedCert.CheckSignatureFrom(issuerCert); err == nil {
+				// Valid mapping. Add it to the specified entry.
+				revInfo.CertificateIssuer = issuerId
```
Comment (kitography) [inline builtin/logical/pki/crl_util.go:641]: nit: Would be great to add a comment here that this isn't stable - that if multiple issuers could have possibly issued this certificate, it's picking the first one in the map.

**Review by kitography — APPROVED** [review]
This looks awesome... and I have nits.

File: builtin/logical/pki/crl_test.go (hunk 2 above)
Comment (cipherboy) [reply builtin/logical/pki/crl_test.go:810]: Ah, it enables the sys/raw endpoint: https://www.vaultproject.io/api/system/raw

File: builtin/logical/pki/crl_test.go (hunk 3 above)
Comment (cipherboy) [reply builtin/logical/pki/crl_test.go:927]: I ended up removing this comment altogether thanks to @stevendpclark's find.

**Pull request merged**